vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js

```diff
@@ -194,7 +194,7 @@ import { ItemCache } from '@vaadin/vaadin-grid/src/vaadin-grid-data-provider-mix
           return;
         }
         if (!newVal) {
-          if (!grid.$connector.deselectAllowed && oldVal) {
+          if (!grid.$connector.deselectAllowed && oldVal && selectedKeys[oldVal.key]) {
```

Comment:
The logic flow could use some cleanup:
```
if (a && b && c) {
  foo();
} else if (b && c) {
  bar();
}
```
->

```
if (b && c) {
  if (a) {
    foo();
  } else {
    bar();
  }
}
```

Reply:
done.